src/v/ssx/metrics.h

```diff
@@ -81,18 +83,105 @@ class metric_groups {
     // Add metrics belonging to the same group.
     //
     // This is the same method as above but using an explicit vector.
-    metric_groups& add_group(
+    metric_groups_base& add_group(
       const ss::metrics::group_name_type& name,
       const std::vector<ss::metrics::metric_definition>& l) {
         _underlying.add_group(name, l);
         return *this;
     }
 
-    // Clear all metric registrations.
-    void clear() { _underlying.clear(); }
+protected:
+    ss::metrics::metric_groups _underlying;
+};
+
+class internal_metric_groups : public metric_groups_base {
+public:
+    internal_metric_groups()
+      : metric_groups_base(ss::metrics::default_handle()) {}
+
+    using metric_groups_base::add_group;
+
+    /*
+     * The two following methods are an extension to the methods above. They
+     * allow registering the metrics in the same way. However when registering
+     * metrics with label aggregation prefer to use this method.
+     *
+     * It allows specifying the aggregation labels to use when the config flag
+     * `aggregate_metrics=false` as well as `aggregate_metrics=true`.
+     *
+     * Further it uses the `metrics_registry` under the hood such that the
+     * aggregation labels will be automatically updated when the config flag
+     * changes.
+     *
+     * The specified labels will be used for all the specified metrics in the
+     * group. If aggregation labels are specified on a per metric basis they
+     * will be overriden.
+     *
+     * Implementation note: Taking `impl::metric_definition_impl` isn't great
+     * but that's a consequence of how the seastar metrics API works.
+     * `ss::metrics::metric_definition` is just a thin wrapper around
+     * `impl::metric_definition_impl` which doesn't expose any of the members
+     * (why?). Further it actually takes a `impl::metric_definition_impl` as its
+     * main constructor argument which is returned by all the `make_` functions.
+     * Hence we are forced to use the `impl` type here as well as we need access
+     * to the metric names.
+     */
+    internal_metric_groups& add_group(
+      const ss::metrics::group_name_type& name,
+      const std::initializer_list<ss::metrics::impl::metric_definition_impl>& l,
```

Comment:
Why do we need this overload with initializer list? Can't the vector one be used with the same syntax?

Reply:
(this would also simplify or remove the need for the private impl function, which I also have some quesitons about but maybe it goes away ....)

Reply:
No, we don't actually need it. Which means we can simplify the implementation a bit.

It was just replicating the original function which has this comment:

```
     * This overload (with initializer_list) is needed because metric_definition
     * has no copy constructor, so the other overload (with vector) cannot be
     * invoked on a braced-init-list.
```

However given we use the impl type here this doesn't apply to us.

Does this make your follow up question go away?

Reply:
Oh right, not being able to use noncopyable objects is an annoying limitation of `std::vector` (and containers in general) which itself derives from their use of std::initializer_list, a bit of a shame. I guess IL and move semantics were both introduced in the same C++ version which sort of lowered the chance of them playing well together due to parallel specification.

Lucky it does not apply to us.

> Does this make your follow up question go away?


Checking the new impl!

Reply:
New impl looks good, my question(s) were obsoleted.